service/kernel/data_management/train_test_split.cpp

```diff
@@ -0,0 +1,253 @@
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "data_management/data/internal/train_test_split.h"
+#include "data_management/data/numeric_table.h"
+#include "data_management/data/data_dictionary.h"
+#include "services/env_detect.h"
+#include "externals/service_dispatch.h"
+#include "service/kernel/service_data_utils.h"
+#include "service/kernel/service_utils.h"
+#include "service/kernel/service_defines.h"
+#include "algorithms/threading/threading.h"
+#include "service/kernel/data_management/service_numeric_table.h"
+#include "data_management/features/defines.h"
+
+namespace daal
+{
+namespace data_management
+{
+namespace internal
+{
+typedef daal::data_management::NumericTable::StorageLayout NTLayout;
+
+const size_t BLOCK_CONST      = 2048;
+const size_t THREADING_BORDER = 8388608;
+
+template <typename Func>
+void runBlocks(const bool inParallel, const size_t nBlocks, Func func)
+{
+    if (inParallel)
+        daal::threader_for(nBlocks, nBlocks, [&](size_t i) { func(i); });
+    else
+        for (size_t i = 0; i < nBlocks; ++i) func(i);
+}
+
+template <typename DataType, typename IdxType>
+services::Status assignColumnValues(const DataType * origDataPtr, const NumericTablePtr dataTable, const IdxType * idxPtr, const size_t startRow,
+                                    const size_t nRows, const size_t iCol)
+{
+    services::Status s;
+
+    BlockDescriptor<DataType> dataBlock;
+    dataTable->getBlockOfColumnValues(iCol, startRow, nRows, readWrite, dataBlock);
+    DataType * dataPtr = dataBlock.getBlockPtr();
+    DAAL_CHECK_MALLOC(dataPtr);
+
+    for (size_t i = 0; i < nRows; ++i) dataPtr[i] = origDataPtr[idxPtr[i]];
+
+    return s;
+}
+
+template <typename DataType, typename IdxType, daal::CpuType cpu>
+services::Status assignColumnSubset(const DataType * origDataPtr, const NumericTablePtr dataTable, const IdxType * idxPtr, const size_t nRows,
+                                    const size_t iCol, const size_t nThreads)
+{
+    services::Status s;
+
+    if (nRows > THREADING_BORDER && nThreads > 1)
+    {
+        const size_t nBlocks = nRows / BLOCK_CONST + !!(nRows % BLOCK_CONST);
+
+        daal::threader_for(nBlocks, nBlocks, [&](size_t iBlock) {
+            const size_t start = iBlock * BLOCK_CONST;
+            const size_t end   = daal::services::internal::min<cpu, size_t>(start + BLOCK_CONST, nRows);
+
+            s |= assignColumnValues<DataType, IdxType>(origDataPtr, dataTable, idxPtr, start, end - start, iCol);
```

Comment:
For parallel regions we need to use thread-safe status: 
```cpp
daal::SafeStatus safeStatus;

daal::threader_for(nBlocks, nBlocks, [&](size_t iBlock) {
    ...
    safe |= ...;
}

return safeStatus.detach();
```

Reply:
Fixed